torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -660,6 +671,48 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
   return std::make_shared<ProcessGroupNCCL::WorkNCCL>(devices);
 }
 
+struct checkFuture {
+  checkFuture(
+      std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work,
+      std::vector<at::Tensor>& outputs,
+      int work_counter,
+      int input_size,
+      std::mutex& mutex)
+      : work_(std::move(work)),
+        outputs_(outputs),
+        work_counter_(work_counter),
+        input_size_(input_size),
+        mutex_(mutex){};
+
+ public:
+  std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work_;
+  std::vector<at::Tensor>& outputs_;
+  bool markFutureCompletedCheck() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    work_counter_ += 1;
+    return work_counter_ == input_size_;
+  };
+  void markFutureCompleted() {
+    at::launch(([this] {
+      work_->getFuture()->markCompleted(torch::jit::IValue(outputs_));
+    }));
+  }
+
+ private:
+  int work_counter_;
+  int input_size_;
+  std::mutex& mutex_;
+};
+
+void cudaStreamCallback(
+    cudaStream_t /* unused */,
+    cudaError_t /* unused */,
+    void* userData) {
+  auto hostfuncdata = (checkFuture*)userData;
```

Comment:
Use `std::dynamic_pointer_cast` and verify the return value of dynamic_pointer_cast is not null.

Reply:
@pritamdamania87 do you mean smth like `dynamic_cast<CheckFutureWork*>((CheckFutureWork*)userData)`. 

This is an issue now. I think I am not properly casting, because even a simple `workCounter_` increment and `std::cout`, prints meaningless stuff. Looks like object is corrupted. Probably because of the way I'm casting it?

Reply:
I meant `auto hostfuncdata = std::dynamic_cast<CheckFutureWork*>(userData)`.

Reply:
@pritamdamania87 I see, the way that I'm doing that cast doesn't provide any protection. Thanks!

Following the discussion in https://stackoverflow.com/questions/55006471/safe-way-to-dynamic-cast-void-to-a-type , I'll just create a base class of CheckFutureWork and `static_cast` to it first and finally do `dynamic_cast` like: 
```
  auto castedUserDataBase = static_cast<CheckFutureWorkBase*>(userData);
  auto castedUserData = dynamic_cast<CheckFutureWork*>(castedUserDataBase);
```
Let me know if a base class is not necessarily needed for our case.